lib/game.rb

```diff
@@ -0,0 +1,90 @@
+class Game
+  attr_reader :rolls
+
+  MAX_PINS = 10
+
+  def initialize
+    @rolls = [] 
+    @max_rolls = 20
+  end
+
+  def roll(pins) 
+    check_pin_number
+    check_game_status
+    @rolls << pins
+      if @rolls.count == 19 && (pins == 10) || @rolls.count == 20 && (@rolls[-1] + @rolls[-2]) == 10
+        @max_rolls += 1
+      elsif first_bowl? && (pins == 10)
+        @rolls << "-" unless completed_all_rolls
+      end
+  end
+
+  def check_game_status
+    raise ("Game Over. Ask for your score!") if completed_all_rolls
+  end
+
+  def total_score
+    pins_score + total_bonus_points
+  end
+  
+  private
+
+  def check_pin_number
+    raise ("This is not valid input.") if pins > MAX_PINS
```

Comment:
The error could be a bit more explicit here, to help the user. Here the user uses the code directly so they probably know what "input" they gave, but this class could be integrated in a bigger system. It's easier for the programmer to track down a bug if the error is more explicit. Something like: "A number between 0 and 10 was expected for the rolls, you entered X".


Which actually makes me realise - if you are going to check the number, you can also check it's positive :)